### ISSUE TITLE
Perforce source control Integration

### ISSUE THREAD

**Issue by Tobias-Nilsson-Sharkmob** [getsentry/sentry#68525]
### Problem Statement

For source control integrations, it seems primarily GIT is supported.
We, and a lot of other software development studios, use Perforce for source control management.

This is not currently available via any integrations in Sentry, which means that we are missing some features like "suggested owners" and source code snippets.

It would be a nice improvement to have a Perforce integration available in Sentry.

### Product Area

Settings - Integrations

**Comment by Dhrumil-Sentry** [comment]
Thanks for this feedback, We'll add it to our backlog and comment here once we know about potential timelines but we're unlikely to tackle this in the near term

**Comment by erebel55** [comment]
This would be wonderful!

**Comment by bruno-garcia** [comment]
Internal link for reference: https://www.notion.so/sentry/11b8b10e4b5d80b9aff9e068237d37b4

**Comment by bruno-garcia** [comment]
[Context from Perforce's State of Gaming 2024](https://www.perforce.com/resources/vcs/game-technology-report)
Download the full report on their website ^

![Image](https://github.com/user-attachments/assets/a7eb9e72-d7ac-4687-8405-26ac0becbe87)

**Comment by bruno-garcia** [comment]
Some of the features we could add:

- Open on Perforce (link to VCS as we have for GitHub)
- Code owners 
- Releases (show changeset per release)

**Comment by zsd4yr** [comment]
Just want to +1 this request

In particular, without this we can’t fully take advantage of release automation and associating p4 changelists with releases

**Comment by bruno-garcia** [comment]
@zsd4yr we haven't being able to prioritize this yet. 
I wonder if community could take this, if we have some documentation on how to add integrations to Sentry

**Comment by zsd4yr** [comment]
@bruno-garcia I fear I lack the language expertise to do so but having docs could certainly help :) Maybe someone else can take it on

**Comment by bruno-garcia** [comment]
Just to record that this came up again in the context of game dev

**Comment by bruno-garcia** [comment]
For reference, an internal doc tracking stuff: https://www.notion.so/sentry/19e8b10e4b5d80eab465c0126f6f36b4

**Comment by bruno-garcia** [comment]
For context, @armenzg shared that InApp frame's are needed to unblock other VCS features such as:

* Stack trace linking (It needs VCS & code mappings)
* Code ownership
* You can have code ownership without CODEOWNERS but you have to manually create the rules*
* Suspect Commits (This would also need VCS & code mappings)

**Comment by bruno-garcia** [comment]
Keeping track of this request here (from a few months ago, but I'll write back once we have something that can be tested): https://sentry.zendesk.com/agent/tickets/135394

**Comment by souredoutlook** [comment]
Listing out some tickets from Perforce Users from the last 2 years:
- https://sentry.zendesk.com/agent/tickets/125365
- https://sentry.zendesk.com/agent/tickets/85487
- https://sentry.zendesk.com/agent/tickets/57937
- https://sentry.zendesk.com/agent/tickets/73108
- https://sentry.zendesk.com/agent/tickets/66493

**Comment by bruno-garcia** [comment]
A customer shared that they use P4 Swarm to define auto reviewers, required reviewers etc:
https://help.perforce.com/helix-core/helix-swarm/swarm/2024.3/Content/Swarm/admin.project_settings.html

Something for us to consider when expanding P4 integration.

**Comment by droberts-krafton** [comment]
At Krafton Europe we also use P4 as our source control for Unreal Engine projects, and also use Swarm for our review process.

To add to the above, since our source control is not integrated, but we do have Jira integration set up, we also see issues when using the "Sync Jira status to Sentry" option. When a bug is marked as fixed in Jira it will then resolve the issue in Sentry, but users internally then may use an older CL or a different branch of P4 that has the issue present and Sentry will then reopen the issue in both Sentry and Jira incorrectly (which we then close manually, potentially multiple times).

To solve this we'd need Perforce integration to tell Sentry "don't open this issue if the CL is higher than X" or "don't open this issue if the branch is X". Please let us know when Perforce integration will be ready 🙏 We're looking forward to it!

**Comment by DeeVine** [comment]
A customer mentioned they're looking forward to the Perforce integration, is there any ETA on this? Thanks!

**Comment by sentaur-athena** [comment]
Unfortunately no timelines yet. I'll update the ticket when we have one.

**Comment by DeeVine** [comment]
hi @sentaur-athena , do we have any ETAs on this yet? Thanks.

**Comment by techwmazur** [comment]
> [Context from Perforce's State of Gaming 2024](https://www.perforce.com/resources/vcs/game-technology-report) Download the full report on their website ^
> 
> ![Image](https://github.com/user-attachments/assets/a7eb9e72-d7ac-4687-8405-26ac0becbe87)

That's more than true (unfortunately). Nevertheless, I know you guys are aiming to support gaming and get even more gaming clients, and to be honest, it might be the missing link. I see that the feature is constantly being dropped, and I would kindly ask you to tackle it more seriously. 

Additionally, Helix has added a review tool (P4 Review) with a rich REST API that should ease up the implementation. You can find more details in the [documentation](https://help.perforce.com/helix-core/helix-swarm/swarm/current/Content/Swarm/swarm-apidoc.html).

**Comment by sentaur-athena** [comment]
Unfortunately there are no timelines yet but we're actively trying to staff it to make it happen in Q4. I'll post a more accurate timeline when we have one.

At this point we're thinking of a gradual launch, enabling one feature at a time. Not all source code integrations features we have can expand to a none-git integrations like Perforce. If you all can help me understand what features of perforce integrations are key to your workflows we can start from there and unblock you faster.

Here is a list of Sentry <> GitHub integration features as reference:

- [Stack trace linking](https://docs.sentry.io/organization/integrations/source-code-mgmt/github/#stack-trace-linking) - go from Sentry issue to your code in github
- Issue assignment - Automatically assign issues based on CODEOWNERS or provided code ownership file
- Get comment on PR when that PR is responsible for an issue on Sentry
- Get comment on PR when that PR is touching files that have open issues on Sentry
- Commits in each release: Sentry releases product provides a list of commits in each realease
- Resolve in commit: Adding fixes <SENTRY-SHORT-ID> to GitHub Pull Request will automatically resolve the sentry issue
- [AI Auto fix](https://docs.sentry.io/product/issues/issue-details/sentry-seer/) is a new Sentry feature. When an issue occurs, auto fix and give users root cause and even make a PR to fix the issue.

**Comment by techwmazur** [comment]
Thanks @sentaur-athena 

Here's a list of key features (with tips for implementation):


###  [Stack trace linking](https://docs.sentry.io/organization/integrations/source-code-mgmt/github/#stack-trace-linking) - go from Sentry issue to your code in github

This one is most important and simplest to achieve see https://help.perforce.com/helix-core/helix-swarm/swarm/current/Content/Swarm/basics.files.html?Highlight=files.

The p4 review supports browsing repo files out of the box.

### Commits in each release: Sentry releases product provides a list of commits in each release

Also very useful feature that can be realized using https://help.perforce.com/helix-core/helix-swarm/swarm/current/Content/Swarm/basics.commits.html?Highlight=commits and range filter,

e.g to list commits between one release and another you can use `@rel1,@rel2`

---

Not a key feature, but still useful

### Get comment on PR when that PR is responsible for an issue on Sentry & Get comment on PR when that PR is touching files that have open issues on Sentry

Can be realized by usage of https://help.perforce.com/helix-core/helix-swarm/swarm/current/Content/Swarm/swarm-apidoc_endpoints-v-current.html review API


---

Nice to have

- Issue assignment - Automatically assign issues based on CODEOWNERS or provided code ownership file
- Resolve in commit: Adding fixes to GitHub Pull Request will automatically resolve the sentry issue
- [AI Auto fix](https://docs.sentry.io/product/issues/issue-details/sentry-seer/) is a new Sentry feature. When an issue occurs, auto fix and give users root cause and even make a PR to fix the issue.

**Comment by sentaur-athena** [comment]
Thank you for the detailed response @techwmazur

**Comment by albertovadia** [comment]
Hi, a bit confused as to why Perforce has not been prioritized. Game developers and other specific industries rely on Perforce for many reasons. Perforce is well integrated and supported by most enterprise solutions. There are studios in the range of 30 to 300 engineers which would gladly switch to Sentry if it supported Perforce. Also, game studios are extremely sticky users because they rely on crash tracking much more than other industries who only bother during early and middle phases.

**Comment by bruno-garcia** [comment]
We've been trying to get this in and **have someone looking at this right now**. That said, we got a doc shared with the Perforce team who basically said a lot of our concepts (from Git) don't map really well to Perforce.
Making this a **very hard problem**. So we'll do our best in adding as many features as we can.

But worth noting that:
You don't need VCS integration for most of Sentry's feature to work. You can get stack traces, source context and even Seer working without VCS. So we hope to make Sentry better through an integration with Perforce but we have thousands of game studios using Sentry successfully without this integration. In fact, most Sentry customers don't have any VCS integration

**Comment by jeremy-wildlight** [comment]
I think the "feature" I would want the most isn't a direct SCM integration actually, but something where I can use to mark an issue as fixed in a CL, and have Sentry not auto-open issues if they are encountered prior to that CL.

I understand this doesn't map to Git though. because being distributed, commit hashes don't increase monotonically, as they do in P4. But I assume this is one of the primary advantages of a Git integration.

If there was an alternative mechanism to gain this capability, we'd use that immediately and not even bother to integrate the product with P4 (in fact, we never would want external vendors to access our depot except in very specific scenarios, even with "no-AI" settings checked -- the danger of information leakage is too high and we will operate on the principle of "least privilege"). You could imagine a custom event tag used for this purpose perhaps?

**Comment by bruno-garcia** [comment]
> I think the "feature" I would want the most isn't a direct SCM integration actually, but something where I can use to mark an issue as fixed in a CL, and have Sentry not auto-open issues if they are encountered prior to that CL.

This is challenging even today with GitHub. This is about Releases. If you version the code with semver, that it's simpler. Sentry understands semver and if you say "Resolved in this release" for an issue on 1.2.3, if that error comes from `release:1.1.1` we won't reopen. If it comes from `release:1.2.4` or `release:2.0.0` it will reopen the issue as a regression. No VCS integration needed. 
Without semver, Sentry basically checks "is this the first time I see this release? no, did it come before or after that one?" so if you have releases coming from different environments, or one doesn't see data for a while, when another does, or release sent by a local dev environment, all that messes things up a bit. We have issues with that with web apps, that use commit sha as the release name, for example. You can improve that by using Sentry CLI to create releases in CI. Then Sentry has a reliable source of "when was this release created" but still not as bullet proof as semver if you have concurrent releases going for example.

> I understand this doesn't map to Git though. because being distributed, commit hashes don't increase monotonically, as they do in P4. But I assume this is one of the primary advantages of a Git integration.


It's not really because of the id. It's a lot of other concepts, like repositories. Pull requests, etc. We have some ideas to create some mapping where you add a part of the depot as the repository manually, for example.
Another is that Perforce isn't 1 thing. They have Helix and Swarm and whatever, different products. And also the fact most customers have a self hosted version of it. So we don't need only to integrate with SaaS but figure out connecting/security/firewalling etc. But that's not the hard part, we support other self hosted VCS. Just doesn't make it easier.

> If there was an alternative mechanism to gain this capability, we'd use that immediately and not even bother to integrate the product with P4 (in fact, we never would want external vendors to access our depot except in very specific scenarios, even with "no-AI" settings checked -- the danger of information leakage is too high and we will operate on the principle of "least privilege"). You could imagine a custom event tag used for this purpose perhaps?

Hopefully the releases note above help. Let me know if there are other areas you need help with.
Releases is [documented here](https://docs.sentry.io/product/releases/). And [Sentry CLI here](https://docs.sentry.io/cli/).

**Comment by jeremy-wildlight** [comment]
As written, my comment isn't quite correct because changelist numbers (in P4) constitute a monotonic timeline within a single stream, but not across streams (I wrote it the way I did because I am likely going to restructure the project to map streams to environments).

> We have issues with that with web apps, that use commit sha as the release name, for example.

Assuming this usage is "fine" from a data modeling perspective, I believe that a "release per CL" would surprisingly almost get something that works, so I'll think about incorporating the CL number directly into a semver-annotated release label.

**Comment by redxdev** [comment]
Throwing my hat in here as my needs are much the same - while true perforce integration would be amazing and I absolutely would want to see, just being able to resolve an issue to a changelist number in order to avoid some amount of "false" regressions from previous changelists would be great. Even if it weren't stream-aware, just using the changelist numbers to ignore older CLs would cut out a bunch of noise from issues falsely regressing.

My team uses changelist numbers as part of our build versioning so we can/do use releases to help with this, but right now you can't resolve to a nonexistent release, Which means we'd either need to wait for the next build (can take hours if not a full day - welcome to gamedev), have a perforce trigger create a new release every time a new change is submitted (seems wasteful), or resolve to "the next release" (results in a false-regression if a build was already in progress without the fix). I actually made a feature request about this previously in https://github.com/getsentry/sentry/issues/83696 as a suggestion for a stopgap until actual perforce integration happens.

**Comment by vaind** [comment]
@jeremy-wildlight what you're trying to achieve should be doable without any source control integration.

As you've already found out - yes, you can just use the CL number directly for the release version. I don't know how you're using streams but assuming a single sequential list of CL numbers that associate with a single project, you can create your semver in multiple ways, the most simple one probably just "1.0.<CL-number>" and use that as the version part of the release. You can create releases for all changelists but also you could just create releases when you actually "release" the game, be it to production, testing, etc - any time you could start getting sentry issues with that CL number. Of course this assumes you associate the release at runtime so when an event is sent to sentry, it has the release number, see https://docs.sentry.io/platforms/native/configuration/options/

**Comment by vaind** [comment]
> My team uses changelist numbers as part of our build versioning so we can/do use releases to help with this, but right now you can't resolve to a nonexistent release,

@redxdev Does using "Resolve in next release" help? See also my previous comments about using CL numbers as semver for releases.

<img width="338" height="171" alt="Image" src="https://github.com/user-attachments/assets/5c097ca6-2cb6-40a4-a4df-c173b41668d9" />

**Comment by jeremy-wildlight** [comment]
> > My team uses changelist numbers as part of our build versioning so we can/do use releases to help with this, but right now you can't resolve to a nonexistent release,
> 
> [@redxdev](https://github.com/redxdev) Does using "Resolve in next release" help? See also my previous comments about using CL numbers as semver for releases.
> 
> <img alt="Image" width="338" height="171" src="https://private-user-images.githubusercontent.com/6349682/501875801-5c097ca6-2cb6-40a4-a4df-c173b41668d9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA2MzE4ODksIm5iZiI6MTc2MDYzMTU4OSwicGF0aCI6Ii82MzQ5NjgyLzUwMTg3NTgwMS01YzA5N2NhNi0yY2I2LTQwYTQtYTRkZi1jMTczYjQxNjY4ZDkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMTZUMTYxOTQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmI4NjcyZTQ0ZDYzODA4MjE5NGU4Mzk0ZGI3ZDdiNjk3Y2FiNjA3ZWNmNmQyZmJkNWMwNjA4YmQ3NmFmYjhjYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.GASI1mJJN660ZwU6FYtsgjIHB82BrsiWa3oYGP98lPo">

I looked at this briefly, and the answer is "no, it doesn't help" unfortunately. Resolving an issue in "the next release" will imply that all events submitted starting on _the next changelist_ with this scheme will reopen the issue immediately. The only way this would work is if the resolution of an event allowed the user to specify the precise release version where the fix is expected (and this would need to be hooked up to Jira integrations or any other issue tracker somehow to synchronize the effect).

**Comment by vaind** [comment]
In case "next release" doesn't help, you could just create the release with the changelist. Having a changelist ID assumes the changes are already committed (AFAIU p4) so at that point you could just submit the release to Sentry, right? It doesn't matter that Sentry won't be seeing any events for a while before the release is rolled out (or ever if it never gets to be published in the wild)

**Comment by srest2021** [comment]
@redxdev @jeremy-wildlight Hi both, we are currently working on a feature that will allow you to resolve in a future, possibly nonexistent release. You would need to specify a semver-formatted version. This way we won't mark an issue as a regression unless it's happening in the release version you specified, or in another semver release greater than that version. I don't think changelist numbers will be in scope as of right now, but a workaround would be to do something like: 1.0.1, 1.0.2, ..., 1.0.28283, etc. Would this help your workflow?

[Linear ticket](https://linear.app/getsentry/project/allow-users-to-resolve-issues-in-specific-possibly-not-yet-created-3182090a5503/overview)

**Comment by vaind** [comment]
@srest2021 thx for pitching in. I can't access the ticket on Linear but from what you're saying it should also close #83696

**Comment by jeremy-wildlight** [comment]
> In case "next release" doesn't help, you could just create the release with the changelist.

@vaind I'm not following I'm afraid.

> CL1000 - bug reported by Sentry, resolved in CL1002 and closed
> CL1001 - bug reported by Sentry again, and reopened as the "next release"
> CL1002 - bug actually fixed in this release.

"Resolved in next release" does not have sufficient granularity to annotate the point in time where a bug is actually resolved.

@srest2021 I can't access that Linear ticket, but yes, being able to use the CL number as the incremental semver value and also mark a specific release version as the "resolved at" version sounds like precisely what would work. Workflow-wise, I think the other remaining improvement beyond this is having the "resolved at version" be sourced from a specific issue label/field in Sentry's Jira integration.

**Comment by redxdev** [comment]
> In case "next release" doesn't help, you could just create the release with the changelist.

I could setup a perforce trigger to create a new release in sentry every time a changelist is submitted but that ends up creating hundreds if not thousands of additional releases that are completely useless and pollute the releases page. It's really not ideal.

And like @jeremy-wildlight said, "resolve in next release" is prone to regressing an issue against a build that didn't have the fix yet -

* Issue is seen in CL#1
* Issue is fixed in CL#3
* A new build on CL#2 is made, issue occurs and sentry marks the issue regressed, despite this build not having the fix.

@srest2021 yes - that is what I originally requested in https://github.com/getsentry/sentry/issues/83696 - we put the changelist number in our versions so it would help as a stopgap measure until there's proper perforce integration (one day, I certainly hope).

**Comment by vaind** [comment]
What I was suggesting as an alternative [here](https://github.com/getsentry/sentry/issues/68525#issuecomment-3411835707) is just creating the release in Sentry, with the expected CL as part of the semver release number. Then you don't need to use "Resolve in next release" but can specify the actual release.

In any case, [the new feature of being able to specify a non-existent semver](https://github.com/getsentry/sentry/issues/68525#issuecomment-3412027782) as a version that will resolve the issue will be even better.

**Comment by jeremy-wildlight** [comment]
@vaind OK I think we were just talking through each other most likely.

> Then you don't need to use "Resolve in next release" but can specify the actual release.

This workflow isn't yet possible because of the issue you mentioned. Manually creating a release for each CL is not a tenable solution for users, and also, not every CL is mapped to a "build" that will ultimately be deployed (and constitute an actual created release thereafter).

**Comment by TheCrossBoy** [comment]
> In case "next release" doesn't help, you could just create the release with the changelist. Having a changelist ID assumes the changes are already committed (AFAIU p4) so at that point you could just submit the release to Sentry, right? It doesn't matter that Sentry won't be seeing any events for a while before the release is rolled out (or ever if it never gets to be published in the wild)

one minor problem with this solution (or not "problem" but annoyance) is that the release monitoring page(s) default time frame is from when the release is _created_, which isn't the same as when it is _released_, which means that if you have i.e. 2 weeks of QA time, monitoring the release is less straightforward because it assumes it is publicly released when you create it

for example crashlytics considers it "released" once either there are 500 unique users or the build has at least 1% of total users (and 2 unique users). this also cleans up the release page because it isn't filled with extra releases

slightly off topic, but relevant to that solution (as it is what we do), so i wanted to mention it

> [@vaind](https://github.com/vaind) OK I think we were just talking through each other most likely.
> 
> > Then you don't need to use "Resolve in next release" but can specify the actual release.
> 
> This workflow isn't yet possible because of the issue you mentioned. Manually creating a release for each CL is not a tenable solution for users, and also, not every CL is mapped to a "build" that will ultimately be deployed (and constitute an actual created release thereafter).

If you use jenkins (or any other build automation), you can have the release be created when the build is made rather than when it is committed. it's what we do using the "webhook" method (found in [your-org].sentry.io/settings/projects/client/release-tracking/)

**Comment by techwmazur** [comment]
Is there any update on that issue?

**Comment by kerenkhatiwada** [comment]
[ZD ticket](https://sentry.zendesk.com/agent/tickets/168075): Adding a +1 from a customer. They’re specifically looking for Perforce support so they can set up code mappings.

**Comment by bruno-garcia** [comment]
Work is quite far along:

Frontend: 
* https://github.com/getsentry/sentry/pull/103172
Backend Integration: 
* https://github.com/getsentry/sentry/pull/103287
Backend Installation:
* https://github.com/getsentry/sentry/pull/103171
Backend Repo/Code-Mapping:
* https://github.com/getsentry/sentry/pull/103635
Backend Stacktrace linking / Commits:
* https://github.com/getsentry/sentry/pull/103636

**Comment by albertovadia** [comment]
Hi, I see the work is very far along, exciting stuff.

Any expectation on how soon we can try it out?

Best,

**Comment by egalev** [comment]
> Work is quite far along:
> 

We have users asking for a Sentry<>Diversion connection. Would you consider integrating [Diversion](https://www.diversion.dev)?

It's an API-first SaaS so hopefully the integration is straightforward: https://docs.diversion.dev/api-reference/developer-guide. We can help with the implementation, do you usually accept external PRs? 

We're also customers of Sentry ourselves so happy to contribute.

**Comment by lustsazeus-lab** [comment]
Perforce support would unlock owner suggestions/snippets for a lot of game studios.

A low-risk path could be: start with read-only metadata sync (changelists -> commits abstraction + file links), then add owner mapping once depot/path conventions are validated. If useful, I can help outline a small connector skeleton + test cases.

**Comment by ReneGreen27** [comment]
Additional request for this feature containing within internal ticket[ #175811](https://sentry.zendesk.com/agent/tickets/175811)
@bruno-garcia May we get an update on this feature release if possible please?

**Comment by mujacica** [comment]
Perforce integration is now in closed beta https://docs.sentry.io/organization/integrations/source-code-mgmt/perforce/ -> will keep the ticket open until the GA, and will close then.

**Comment by mujacica** [comment]
Perforce integration is GA now: https://blog.sentry.io/perforce-integration-ga/
